pkg/plugin/plugin_test.go

```diff
@@ -29,99 +33,251 @@ var testHandshake = goPlugin.HandshakeConfig{
 }
 
-func TestRunSuccessfully(t *testing.T) {
-	utils.InitLogger(slog.LevelDebug)
+var _ = Describe("Test TrafficRouter plugin for OpenShift route", func() {
+	var (
+		ctx         context.Context
+		cancel      context.CancelFunc
+		closeCh     chan struct{}
+		routePlugin rolloutsPlugin.TrafficRouterPlugin
+		fakeClient  *fake.Clientset
+	)
+	BeforeEach(func() {
+		utils.InitLogger(slog.LevelDebug)
 
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+		ctx, cancel = context.WithCancel(context.Background())
 
-	s := runtime.NewScheme()
+		s := runtime.NewScheme()
+		Expect(routev1.AddToScheme(s)).To(BeNil())
 
-	if err := routev1.AddToScheme(s); err != nil {
-		t.Fatal("unable to add scheme")
-	}
+		fakeClient = fake.NewSimpleClientset(mocks.MakeObjects()...)
+		routePluginImp := &RpcPlugin{
+			routeClient: fakeClient,
+		}
 
-	dynClient := fakeDynClient.NewSimpleDynamicClient(s, mocks.MakeObjects()...)
-	rpcPluginImp := &RpcPlugin{
-		IsTest:        true,
-		dynamicClient: dynClient,
-	}
+		// pluginMap is the map of plugins we can dispense.
+		pluginName := "RpcTrafficRouterPlugin"
+		var pluginMap = map[string]goPlugin.Plugin{
+			pluginName: &rolloutsPlugin.RpcTrafficRouterPlugin{Impl: routePluginImp},
+		}
 
-	// pluginMap is the map of plugins we can dispense.
-	var pluginMap = map[string]goPlugin.Plugin{
-		"RpcTrafficRouterPlugin": &rolloutsPlugin.RpcTrafficRouterPlugin{Impl: rpcPluginImp},
-	}
+		ch := make(chan *goPlugin.ReattachConfig, 1)
+		closeCh = make(chan struct{})
+		go goPlugin.Serve(&goPlugin.ServeConfig{
+			HandshakeConfig: testHandshake,
+			Plugins:         pluginMap,
+			Test: &goPlugin.ServeTestConfig{
+				Context:          ctx,
+				ReattachConfigCh: ch,
+				CloseCh:          closeCh,
+			},
+		})
 
-	ch := make(chan *goPlugin.ReattachConfig, 1)
-	closeCh := make(chan struct{})
-	go goPlugin.Serve(&goPlugin.ServeConfig{
-		HandshakeConfig: testHandshake,
-		Plugins:         pluginMap,
-		Test: &goPlugin.ServeTestConfig{
-			Context:          ctx,
-			ReattachConfigCh: ch,
-			CloseCh:          closeCh,
-		},
-	})
+		// We should get a config
+		var config *goPlugin.ReattachConfig
+		select {
+		case config = <-ch:
+		case <-time.After(2000 * time.Millisecond):
+			Fail("should've received reattach")
+		}
 
-	// We should get a config
-	var config *goPlugin.ReattachConfig
-	select {
-	case config = <-ch:
-	case <-time.After(2000 * time.Millisecond):
-		t.Fatal("should've received reattach")
-	}
-	if config == nil {
-		t.Fatal("config should not be nil")
-	}
+		Expect(config).ToNot(BeNil())
+
+		// Connect!
+		c := goPlugin.NewClient(&goPlugin.ClientConfig{
+			Cmd:             nil,
+			HandshakeConfig: testHandshake,
+			Plugins:         pluginMap,
+			Reattach:        config,
+		})
+		client, err := c.Client()
+		Expect(err).To(BeNil())
+
+		// Pinging should work
+		err = client.Ping()
+		Expect(err).ToNot(HaveOccurred())
+
+		// Kill which should do nothing
+		c.Kill()
```

Comment:
Looks like it got added from the [rollouts-repo](https://github.com/argoproj/argo-rollouts/blob/36c8009cfc9805837573ff58d148c06ce2c078e3/test/cmd/metrics-plugin-sample/internal/plugin/plugin_test.go#L81).  I think it was included to verify if the plugin is configured in test mode. Since we are using it in a unit test it shouldn't do anything. But in production, it is expected to kill the plugin subprocess. 

@ishitasequeira Please feel free to correct me.

Reply:
@chetan-rns you described it correctly. The kill basically kills the plugin process and makes sure that no pings are sent to it when the plugin is not running.